### PR TITLE
docs: adapt doc generation to match tooling expected format

### PR DESCRIPTION
**Pull request by gssbzn**
## Proposed changes

I have adapted the way cobra generates docs to match the expected format of our tooling and style

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run `make fmt` and formatted my code

## Further comments

For a sample of the generate files download [this](https://mongodb-mongocli-build.s3.amazonaws.com/mongocli-master/docs/cd2c338e8e6742ea701c1240a916f454c53cc2f7_21_04_15_19_25_54/reference.tgz)

To see how this renders see [here](https://docs-mongodbcom-staging.corp.mongodb.com/mongocli/docsworker-xlarge/gen-docs-test/reference/mongocli-atlas/)

### CODE REVIEW

File: build/ci/evergreen.yml [hunk 1]
```diff
@@ -992,7 +979,22 @@ tasks:
     commands:
       - func: "clone"
       - func: "generate docs"
-      - func: "upload docs"
+      - command: archive.targz_pack
```
Comment (gssbzn) [inline build/ci/evergreen.yml:982]: targz all the docs and upload them as a single file to make the downloading of the whole docs easier

File: internal/docs/rest/rest.go [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+func (s byName) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
+func (s byName) Less(i, j int) bool { return s[i].Name() < s[j].Name() }
+
+func printOptionsReST(buf *bytes.Buffer, cmd *cobra.Command) error {
+	flags := cmd.NonInheritedFlags()
+	if flags.HasAvailableFlags() {
+		buf.WriteString("Options\n")
+		buf.WriteString("~~~~~~~\n\n.. list-table::\n")
+		buf.WriteString(`   :header-rows: 1
+   :widths: 30 20 50
```
Comment (melissamahoney-mongodb) [inline internal/docs/rest/rest.go:166]: This was my fault for missing this, need to add a width for the new fourth column
```suggestion
   :widths: 30 10 50 10
```

Comment (melissamahoney-mongodb) [inline internal/docs/rest/rest.go]:
```suggestion
   :widths: 20 10 60 10
```
This is my fault, I forgot to specify a width for the new column.

File: internal/docs/rest/rest.go [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+		buf.WriteString(indentString(FlagUsages(flags), " "))
+		buf.WriteString("\n")
+	}
+
+	parentFlags := cmd.InheritedFlags()
+	if parentFlags.HasAvailableFlags() {
+		buf.WriteString("Inherited Options\n")
+		buf.WriteString("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n\n.. list-table::\n")
+		buf.WriteString(`   :header-rows: 1
+   :widths: 30 20 50
```
Comment (melissamahoney-mongodb) [inline internal/docs/rest/rest.go]:
```suggestion
   :widths: 20 10 60 10
```

File: internal/docs/rest/rest.go [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+	}
+
+	if len(cmd.Example) > 0 {
+		buf.WriteString("Examples\n")
+		buf.WriteString("~~~~~~~~\n\n")
+		buf.WriteString(fmt.Sprintf(".. code-block::\n%s\n\n", indentString(cmd.Example, " ")))
+	}
+
+	if hasSeeAlso(cmd) {
+		buf.WriteString("SEE ALSO\n")
```
Comment (melissamahoney-mongodb) [inline internal/docs/rest/rest.go:104]:
```suggestion
		buf.WriteString("See Also\n")
```

File: internal/cli/atlas/clusters/create.go [hunk 5]
```diff
@@ -232,7 +232,12 @@ Some of the cluster configuration options are available via flags but for full c
 		RunE: func(cmd *cobra.Command, args []string) error {
 			return opts.Run()
 		},
+		Annotations: map[string]string{
+			"args":     "name",
```
Comment (gssbzn) [inline internal/cli/atlas/clusters/create.go:236]: how to get the args and descriptions from a command, we'll need to back fill this

File: internal/docs/rest/rest.go [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+		buf.WriteString(optionsHeader)
+		var requiredSlice []string
+		if requiredArgs, hasRequired := cmd.Annotations["requiredArgs"]; hasRequired {
+			requiredSlice = strings.Split(requiredArgs, ",")
+		}
+
+		for _, arg := range strings.Split(args, ",") {
+			required := search.StringInSlice(requiredSlice, arg)
+			description := cmd.Annotations[arg+"Desc"]
+			line := fmt.Sprintf("   * - %s\n     - %v\n     - %s", arg, required, description)
```
Comment (melissamahoney-mongodb) [inline internal/docs/rest/rest.go:206]: I think we're missing the "type" column here.

File: internal/docs/rest/rest.go [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+
+func (s byName) Len() int           { return len(s) }
+func (s byName) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
+func (s byName) Less(i, j int) bool { return s[i].Name() < s[j].Name() }
+
+const optionsHeader = `.. list-table::
+   :header-rows: 1
+   :widths: 20 10 10 60
+
+   * - Option
```
Comment (melissamahoney-mongodb) [inline internal/docs/rest/rest.go:187]: If we're going to reuse this for all three tables, we'll need to think of a more generic word than "option", since that won't work for the "argument" section. Possibly "Field"? or just "Name"?

File: internal/docs/rest/rest.go (hunk 6 above)
Comment (gssbzn) [reply internal/docs/rest/rest.go:206]: args don't have type, they're always string, not sure how you may work with that

Comment (melissamahoney-mongodb) [reply internal/docs/rest/rest.go:232]: Can we just put in string for everything then? to make the tables consistent?

File: internal/docs/rest/rest.go (hunk 7 above)
Comment (gssbzn) [reply internal/docs/rest/rest.go]: I'm reusing because I'm lazy if you need them to be different feel free to suggest so

File: internal/docs/rest/rest.go (hunk 6 above)
Comment (gssbzn) [reply internal/docs/rest/rest.go]: sure

File: internal/cli/atlas/atlas.go [hunk 8]
```diff
@@ -58,6 +58,9 @@ func Builder() *cobra.Command {
 
 			return validate.Credentials()
 		},
+		Annotations: map[string]string{
+			"toc": "true",
```
Comment (gssbzn) [inline internal/cli/atlas/atlas.go:62]: to mark doc to generate the Table of content (ToC) for this file

File: internal/docs/rest/rest.go [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+			}
+			cname := name + " " + child.Name()
+			ref = strings.ReplaceAll(cname, " ", separator)
+			buf.WriteString(fmt.Sprintf("   /reference/%s\n", ref))
+		}
+		buf.WriteString("\n")
+	}
+
+	if !cmd.DisableAutoGenTag {
+		buf.WriteString("*Auto generated by MongoDB CLI on " + time.Now().Format("2-Jan-2006") + "*\n")
```
Comment (colm-quinn) [inline internal/docs/rest/rest.go:159]: go and their magical reference dates... :P

**Review by colm-quinn — DISMISSED** [review]
Looks great, LGTM. I'll let docs have the final say.

As an aside, it would be useful to include a description of the data that's generated by `rest.go`, as the final docs link currently doesn't show the intermediate format that these changes are generating and sending to tools.

File: internal/docs/rest/rest.go [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+		children := cmd.Commands()
+		sort.Sort(byName(children))
+
+		for _, child := range children {
+			if !child.IsAvailableCommand() || child.IsAdditionalHelpTopicCommand() {
+				continue
+			}
+			cname := name + " " + child.Name()
+			ref = strings.ReplaceAll(cname, " ", separator)
+			buf.WriteString(fmt.Sprintf("   /reference/%s\n", ref))
```
Comment (melissamahoney-mongodb) [inline internal/docs/rest/rest.go:153]: Would it be possible to produce a list of child commands like this?
```
create </reference/mongocli-atlas-accessLists-create>
delete </reference/mongocli-atlas-accessLists-delete>
describe </reference/mongocli-atlas-accessLists-describe>
list </reference/mongocli-atlas-accessLists-list>
```
This would greatly reduce clutter in the left nav. See a [staged example](https://docs-mongodbcom-staging.corp.mongodb.com/mongocli/docsworker-xlarge/gen-docs-test/reference/mongocli-atlas-accessLists/).
I realize this is a late request so we don't have to do it in this PR, just wondering if its possible!

**Review by melissamahoney-mongodb — DISMISSED** [review]
Thank you SO MUCH for all your efforts here!

**Comment by gssbzn** [comment]
@andreaangiolillo do you have any suggestions here?

File: build/ci/ego [hunk 11]
```diff
@@ -17,8 +17,10 @@
 set -o nounset
 set -o errexit
 
-readonly DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
-readonly SCRIPT=$(basename "${BASH_SOURCE[0]}")
+readonly DIR=
```
Comment (gssbzn) [inline build/ci/ego:20]: I think the llinter for this got updated and it now complains, happy to make a separate PR

File: internal/docs/rest/rest.go (hunk 10 above)
Comment (gssbzn) [reply internal/docs/rest/rest.go]: done

**Review by fmenezes — APPROVED** [review]
Restamp

**Review by melissamahoney-mongodb — APPROVED** [review]
The change for the left nav looks good! Thank you!
